File: app/src/main/java/com/jobtick/activities/NewTaskAlertsActivity.java
package com.jobtick.activities;

import android.app.Dialog;
import android.content.Intent;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.os.Bundle;
import android.view.MenuItem;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.widget.RadioButton;
import android.widget.RadioGroup;

import androidx.appcompat.widget.AppCompatButton;
import androidx.viewpager.widget.ViewPager;

import com.android.volley.AuthFailureError;
import com.android.volley.DefaultRetryPolicy;
import com.android.volley.NetworkResponse;
import com.android.volley.RequestQueue;
import com.android.volley.toolbox.StringRequest;
import com.android.volley.toolbox.Volley;
import com.google.android.material.appbar.MaterialToolbar;
import com.jobtick.R;
import android.annotation.SuppressLint;

import com.jobtick.text_view.TextViewRegular;
import com.jobtick.adapers.SectionsPagerAdapter;
import com.jobtick.fragments.NewTaskAlertsInPersonFragment;
import com.jobtick.fragments.NewTaskAlertsRemoteFragment;
import com.jobtick.models.task.TaskAlert;
import com.jobtick.utils.Constant;
import com.jobtick.utils.HttpStatus;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.HashMap;
import java.util.Map;

import butterknife.BindView;
import butterknife.ButterKnife;
import timber.log.Timber;

public class NewTaskAlertsActivity extends ActivityBase implements NewTaskAlertsInPersonFragment.OperationInPersonListener, NewTaskAlertsRemoteFragment.OperationRemoteListener {

    @SuppressLint("NonConstantResourceId")
    @BindView(R.id.toolbar)
    MaterialToolbar toolbar;
    @SuppressLint("NonConstantResourceId")
    @BindView(R.id.rb_in_person)
    RadioButton rbInPerson;
    @SuppressLint("NonConstantResourceId")
    @BindView(R.id.rb_remote)
    RadioButton rbRemote;
    @SuppressLint("NonConstantResourceId")
    @BindView(R.id.rg_inPerson_remote)
    RadioGroup rgInPersonRemote;
    @SuppressLint("NonConstantResourceId")
    @BindView(R.id.view_pager)
    ViewPager viewPager;

    TaskAlert taskAlert;
    int position = -1;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_new_task_alerts);
        ButterKnife.bind(this);
        taskAlert = new TaskAlert();
        if (getIntent().getExtras() != null) {
            taskAlert = (TaskAlert) getIntent().getExtras().getParcelable("TASK_ALERT");
            position = getIntent().getExtras().getInt("POSITION");
        }
        initToolbar();
        initComponent();
    }

    private void initComponent() {
        setupViewPager(viewPager);
        viewPager.addOnPageChangeListener(viewPagerPageChangeListener);
        if (taskAlert.isValid()) {
            if (taskAlert.getAlert_type().equalsIgnoreCase("remote")) {
                viewPager.setCurrentItem(1);
                rbInPerson.setChecked(false);
                rbRemote.setChecked(true);
            } else {
                viewPager.setCurrentItem(0);
                rbInPerson.setChecked(true);
                rbRemote.setChecked(false);
            }
        } else {
            viewPager.setCurrentItem(0);
        }
        viewPager.setOffscreenPageLimit(2);
        rgInPersonRemote.setOnCheckedChangeListener(new RadioGroup.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(RadioGroup group, int checkedId) {
                switch (group.findViewById(checkedId).getId()) {
                    case R.id.rb_in_person:
                        viewPager.setCurrentItem(0);
                        break;
                    case R.id.rb_remote:
                        viewPager.setCurrentItem(1);
                        break;
                }
            }
        });
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                onBackPressed();
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onBackPressed() {
        if (taskAlert.isValid()) {
            Intent intent = new Intent();
            Bundle bundle = new Bundle();
            bundle.putInt("POSITION", position);
            bundle.putParcelable("TASK_ALERT", taskAlert);
            intent.putExtras(bundle);
            setResult(1, intent);
        }
        super.onBackPressed();
    }

    private void setupViewPager(ViewPager viewPager) {
        SectionsPagerAdapter adapter = new SectionsPagerAdapter(getSupportFragmentManager());
        adapter.addFragment(NewTaskAlertsInPersonFragment.newInstance(taskAlert, position, this), "New Job Alert");
        adapter.addFragment(NewTaskAlertsRemoteFragment.newInstance(taskAlert, position, this), "New Job Alert");
        viewPager.setAdapter(adapter);
    }


    private void initToolbar() {
        toolbar.setNavigationIcon(R.drawable.ic_back);
        setSupportActionBar(toolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setTitle("New Job Alert");
    }

    //  viewpager change listener
    ViewPager.OnPageChangeListener viewPagerPageChangeListener = new ViewPager.OnPageChangeListener() {

        @Override
        public void onPageSelected(final int position) {
            switch (position) {
                case 0:
                    rbInPerson.setChecked(true);
                    rbRemote.setChecked(false);
                    break;
                case 1:
                    rbInPerson.setChecked(false);
                    rbRemote.setChecked(true);
                    break;
            }

        }

        @Override
        public void onPageScrolled(int arg0, float arg1, int arg2) {

        }

        @Override
        public void onPageScrollStateChanged(int arg0) {

        }
    };


    @Override
    public void onInPersonSave(int position, TaskAlert taskAlert) {
        this.taskAlert = taskAlert;
        this.position = position;
        addTaskAlert(this.taskAlert);
    }

    @Override
    public void onRemoteSave(int position, TaskAlert taskAlert) {
        this.taskAlert = taskAlert;
        this.position = position;
        addTaskAlert(this.taskAlert);

    }


    private void addTaskAlert(TaskAlert taskAlert) {
        //{{baseurl}}/taskalerts
        showProgressDialog();
        StringRequest stringRequest = new StringRequest(StringRequest.Method.POST, Constant.URL_TASK_ALERT,
                response -> {
                    Timber.e(response);
                    hideProgressDialog();
                    try {
                        JSONObject jsonObject = new JSONObject(response);
                        Timber.e(jsonObject.toString());
                        if (jsonObject.has("success") && !jsonObject.isNull("success")) {
                            if (jsonObject.getBoolean("success")) {
                                if (jsonObject.has("data") && !jsonObject.isNull("data")) {
                                    finish();
                                }
                            } else {
                                showToast("Something went Wrong", NewTaskAlertsActivity.this);
                            }
                        }
                    } catch (JSONException e) {
                        Timber.e(String.valueOf(e));
                        e.printStackTrace();

                    }


                },
                error -> {
                    NetworkResponse networkResponse = error.networkResponse;
                    if (networkResponse != null && networkResponse.data != null) {
                        String jsonError = new String(networkResponse.data);
                        // Print Error!
                        Timber.e(jsonError);
                        if (networkResponse.statusCode == HttpStatus.AUTH_FAILED) {
                            unauthorizedUser();
                            hideProgressDialog();
                            return;
                        }
                        try {
                            hideProgressDialog();
                            JSONObject jsonObject = new JSONObject(jsonError);
                            JSONObject jsonObject_error = jsonObject.getJSONObject("error");
                            showToast(jsonObject_error.getString("message"), this);

                        } catch (JSONException e) {
                            e.printStackTrace();
                        }
                    } else {
                        showToast("Something Went Wrong", NewTaskAlertsActivity.this);
                    }
                    Timber.e(error.toString());
                    hideProgressDialog();
                }) {


            @Override
            public Map<String, String> getHeaders() {
                Map<String, String> map1 = new HashMap<String, String>();
                map1.put("authorization", sessionManager.getTokenType() + " " + sessionManager.getAccessToken());
                map1.put("Content-Type", "application/x-www-form-urlencoded");
                map1.put("X-Requested-With", "XMLHttpRequest");
                return map1;
            }

            @Override
            protected Map<String, String> getParams() throws AuthFailureError {
                Map<String, String> map1 = new HashMap<String, String>();
                map1.put("text_phrase", taskAlert.getKetword());
                map1.put("task_type", taskAlert.getAlert_type());
                if (taskAlert.getAlert_type().equals("physical")) {
                    map1.put("location", taskAlert.getSuburb());
                    map1.put("latitude", taskAlert.getLattitude().toString());
                    map1.put("longitude", taskAlert.getLongitude().toString());
                    map1.put("distance", String.valueOf(taskAlert.getDistance()));
                }
                return map1;
            }
        };

        stringRequest.setRetryPolicy(new DefaultRetryPolicy(0, -1,
                DefaultRetryPolicy.DEFAULT_BACKOFF_MULT));
        RequestQueue requestQueue = Volley.newRequestQueue(NewTaskAlertsActivity.this);
        requestQueue.add(stringRequest);
        Timber.e(stringRequest.getUrl());
    }
}

File: app/src/main/java/com/jobtick/android/activities/NewTaskAlertsActivity.kt
package com.jobtick.android.activities

import android.content.Intent
import android.os.Bundle
import android.os.Parcelable
import android.view.MenuItem
import com.android.volley.*
import com.android.volley.toolbox.StringRequest
import com.android.volley.toolbox.Volley
import com.google.android.material.appbar.MaterialToolbar
import com.jobtick.android.BuildConfig
import com.jobtick.android.R
import com.jobtick.android.fragments.NewTaskAlertsInPersonFragment.Companion.newInstance
import com.jobtick.android.fragments.NewTaskAlertsInPersonFragment.OperationInPersonListener
import com.jobtick.android.models.task.TaskAlert
import com.jobtick.android.utils.Constant
import com.jobtick.android.utils.HttpStatus
import org.json.JSONException
import org.json.JSONObject
import timber.log.Timber
import java.util.*

class NewTaskAlertsActivity : ActivityBase(), OperationInPersonListener {

    var toolbar: MaterialToolbar? = null

    var taskAlert: TaskAlert? = null
    var position = -1
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_new_task_alerts)
        toolbar = findViewById(R.id.toolbar)
        taskAlert = TaskAlert()
        if (intent.extras != null) {
            taskAlert = intent.extras!!.getParcelable<Parcelable>("TASK_ALERT") as TaskAlert?
            position = intent.extras!!.getInt("POSITION")
        }
        initToolbar()
        if (savedInstanceState == null) {
            supportFragmentManager.beginTransaction()
                    .setReorderingAllowed(true)
                    .add(R.id.fragment_container_view, newInstance(taskAlert, position, this), "New Job Alert")
                    .commit()
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> onBackPressed()
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onBackPressed() {
        if (taskAlert!!.isValid) {
            val intent = Intent()
            val bundle = Bundle()
            bundle.putInt("POSITION", position)
            bundle.putParcelable("TASK_ALERT", taskAlert)
            intent.putExtras(bundle)
            setResult(1, intent)
        }
        super.onBackPressed()
    }

    private fun initToolbar() {
        toolbar!!.setNavigationIcon(R.drawable.ic_back)
        setSupportActionBar(toolbar)
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)
        supportActionBar!!.title = "New Job Alert"
    }

    override fun onInPersonSave(position: Int, taskAlert: TaskAlert?, minPrice: String, maxPrice: String) {
        this.taskAlert = taskAlert
        this.position = position
        if (taskAlert!!.alert_type == "physical") {
            taskAlert.lattitude = taskAlert.lattitude
            taskAlert.longitude = taskAlert.longitude
            taskAlert.suburb = taskAlert.suburb
        } else {
            taskAlert.lattitude = null
            taskAlert.longitude = null
            taskAlert.suburb = null
        }
        addTaskAlert(this.taskAlert, minPrice, maxPrice)
    }

    private fun addTaskAlert(taskAlert: TaskAlert?, minPrice: String, maxPrice: String) {
        //{{baseurl}}/taskalerts
        showProgressDialog()
        val stringRequest: StringRequest = object : StringRequest(Method.POST, Constant.URL_TASK_ALERT_V2,
                Response.Listener { response: String? ->
                    Timber.e(response)
                    hideProgressDialog()
                    try {
                        val jsonObject = JSONObject(response)
                        Timber.e(jsonObject.toString())
                        if (jsonObject.has("success") && !jsonObject.isNull("success")) {
                            if (jsonObject.getBoolean("success")) {
                                if (jsonObject.has("data") && !jsonObject.isNull("data")) {
                                    finish()
                                }
                            } else {
                                showToast("Something went Wrong", this@NewTaskAlertsActivity)
                            }
                        }
                    } catch (e: JSONException) {
                        Timber.e(e.toString())
                        e.printStackTrace()
                    }
                },
                Response.ErrorListener { error: VolleyError ->
                    val networkResponse = error.networkResponse
                    if (networkResponse != null && networkResponse.data != null) {
                        val jsonError = String(networkResponse.data)
                        // Print Error!
                        Timber.e(jsonError)
                        if (networkResponse.statusCode == HttpStatus.AUTH_FAILED) {
                            unauthorizedUser()
                            hideProgressDialog()
                            return@ErrorListener
                        }
                        try {
                            hideProgressDialog()
                            val jsonObject = JSONObject(jsonError)
                            val jsonObject_error = jsonObject.getJSONObject("error")
                            showToast(jsonObject_error.getString("message"), this)
                        } catch (e: JSONException) {
                            e.printStackTrace()
                        }
                    } else {
                        showToast("Something Went Wrong", this@NewTaskAlertsActivity)
                    }
                    Timber.e(error.toString())
                    hideProgressDialog()
                }) {
            override fun getHeaders(): Map<String, String> {
                val map1: MutableMap<String, String> = HashMap()
                map1["authorization"] = sessionManager.tokenType + " " + sessionManager.accessToken
                map1["Content-Type"] = "application/x-www-form-urlencoded"
                map1["X-Requested-With"] = "XMLHttpRequest"
                map1["Version"] = BuildConfig.VERSION_CODE.toString()
                return map1
            }

            @Throws(AuthFailureError::class)
            override fun getParams(): Map<String, String> {
                val map1: MutableMap<String, String> = HashMap()
                map1["keyword"] = taskAlert!!.ketword
                map1["type"] = taskAlert.alert_type
                map1["minprice"] = minPrice
                map1["maxprice"] = maxPrice
                if (taskAlert.alert_type == "physical") {
                    map1["location"] = taskAlert.suburb
                    map1["latitude"] = taskAlert.lattitude.toString()
                    map1["longitude"] = taskAlert.longitude.toString()
                    map1["distance"] = taskAlert.distance.toString()
                }
                return map1
            }
        }
        stringRequest.retryPolicy = DefaultRetryPolicy(0, -1,
                DefaultRetryPolicy.DEFAULT_BACKOFF_MULT)
        val requestQueue = Volley.newRequestQueue(this@NewTaskAlertsActivity)
        requestQueue.add(stringRequest)
        Timber.e(stringRequest.url)
    }
}